### ISSUE TITLE
Better Reverb

### ISSUE THREAD

**Issue by dungxnd** [dungxnd/ViPERDSP#9]
## 📄 Paper Overview

**Full Title:** Optimizing tiny colorless feedback delay networks

**Authors:** Gloria Dal Santo, Karolina Prawda, Sebastian J. Schlecht, Vesa Välimäki

**Publication:** EURASIP Journal on Audio, Speech, and Music Processing (2025); earlier version presented at DAFx23

**ArXiv ID:** 2402.11216

**Core Innovation:** An optimization framework that uses a **differentiable feedback delay network (FDN)** with as few as **4 delay lines** to learn parameters that produce colorless (smooth, non-metallic) reverberation.

---

### What is "Coloration"?

Artificial reverberation algorithms often suffer from **spectral coloration** — audible **metallic ringing** that degrades perceived sound quality. This is particularly problematic when **fewer delay lines** are used (for computational efficiency), as coloration becomes more pronounced.

### Why is this a hard problem?

In an FDN, coloration is caused by an **uneven distribution of modal excitation** — certain resonant modes are excited more strongly than others, and these strong modes are perceived as metallic ringing. The modal excitation depends on *all* FDN parameters (feedback matrix, input gains, output gains), making it difficult to optimize manually.

---

## 🧠 The Core Idea: Differentiable FDN (DiffFDN)

The paper's key insight is to make the FDN **differentiable**, allowing the use of **gradient-based optimization** (backpropagation) to learn parameters that minimize coloration.

### What makes it "Differentiable"?

The FDN is implemented as a computational graph where all operations (delay lines, matrix multiplication, gain stages) are differentiable. This means:
1. You can compute the **gradient** of a loss function with respect to every parameter.
2. You can use **gradient descent** to iteratively update parameters to minimize the loss.
3. The optimization runs **offline** (in Python/PyTorch) — the resulting parameters are then hardcoded into your C++ implementation.

### Architecture of a Basic FDN

A standard FDN consists of:

```
Input → [Input Gains] → [Delay Lines] → [Feedback Matrix] → [Output Gains] → Output
                              ↑                                |
                              └────────────────────────────────┘
```

- **N delay lines** (the paper uses N=4)
- **Input gains** (b): N coefficients controlling how much signal enters each delay line
- **Feedback matrix** (A): N×N matrix mixing the outputs of delay lines back to their inputs
- **Output gains** (c): N coefficients controlling the contribution of each delay line to the output

---

## 🎯 Optimization Objective (The "Loss Function")

The optimization has a **twofold objective**:

### 1. Spectral Loss (Maximize Spectral Flatness)

**Goal:** Achieve a flat magnitude response (colorless sound).

**How:** Minimize the variance of the FDN's magnitude response across frequency.

**Why:** A flat frequency response means no frequencies are overly emphasized, eliminating the "metallic ringing" caused by strong resonant modes.

### 2. Temporal Loss (Maintain Density)

**Goal:** Prevent the impulse response from becoming too sparse (which would sound like discrete echoes).

**How:** Penalize sparseness in the impulse response.

**Why:** An FDN with a perfectly flat magnitude response can suffer from **temporal buildup of echoes** — the sound becomes too "clean" and lacks the dense, diffuse quality of natural reverberation. The temporal loss ensures the impulse response remains **dense** while still being **colorless**.

### The Trade-off

These two objectives are in tension:
- Maximizing spectral flatness → narrower modal excitation distribution
- Maintaining temporal density → prevents sparse impulse responses

The optimization finds the **sweet spot** where modal excitation is narrowly distributed (no strong ringing modes) while impulse response density is preserved.

---

### What are "Modes"?

In an FDN, **modes** are the resonant frequencies of the system. Each mode has:
- A **frequency** (where it resonates)
- An **excitation level** (how strongly it's excited)

### The Problem

When excitation levels are **widely distributed**, some modes are much stronger than others. These strong modes create the perception of **metallic ringing**.

### The Solution

The optimization aims for a **narrow distribution** of modal excitation — all modes are excited at roughly the same level, so no single frequency rings out.

---

### Phase 1: Understand the FDN Architecture

Study the basic FDN structure:

```
y(n) = c^T * z(n)
z(n+1) = A * z(n) + b * x(n)
```

Where:
- `x(n)` is the input sample
- `z(n)` is the N×1 state vector (delay line contents)
- `A` is the N×N feedback matrix
- `b` is the N×1 input gain vector
- `c` is the N×1 output gain vector
- `y(n)` is the output sample

### Phase 2: Obtain Optimized Parameters

**You do NOT need to run the optimization yourself.** The paper provides:

1. **Pre-computed parameters** for various configurations (N=4, 8, 16)
2. **Companion code** on GitHub (`gdalsanto/diff-fdn-colorless`)
3. **Audio demos** at the companion website

**For ViperDSP:** You can take the optimized parameters from the paper/github and hardcode them into your C++ implementation.

### Phase 3: Implement the FDN in C++

```cpp
class ColorlessFDN {
private:
    static constexpr int N = 4;  // Number of delay lines
    
    // Optimized parameters (from paper)
    double A[N][N];  // Feedback matrix
    double b[N];     // Input gains
    double c[N];     // Output gains
    
    double z[N];     // Delay line state (history)
    double delays[N]; // Delay lengths in samples
    
public:
    double Process(double input) {
        // 1. Compute output
        double output = 0.0;
        for (int i = 0; i < N; i++) {
            output += c[i] * z[i];
        }
        
        // 2. Update delay lines (with feedback)
        double next_z[N];
        for (int i = 0; i < N; i++) {
            double sum = b[i] * input;
            for (int j = 0; j < N; j++) {
                sum += A[i][j] * z[j];
            }
            next_z[i] = sum;
        }
        
        // 3. Shift delay lines (store for next sample)
        for (int i = 0; i < N; i++) {
            z[i] = next_z[i];
        }
        
        return output;
    }
};
```

### Phase 4: Add Attenuation Filters (For Room Impulse Responses)

The paper demonstrates two application scenarios:

1. **Synthetic Room Impulse Responses:** Add attenuation filters to control frequency-dependent decay (high frequencies decay faster than low frequencies, simulating real rooms).

2. **Optimizable Scattering:** Use the optimized feedback matrix to create a rich, diffuse sound field.

For ViperDSP, you may want to implement:
- **Option A:** A simple, colorless reverb (just the optimized FDN)
- **Option B:** A room simulator (FDN + attenuation filters)

---

### Key Advantage: Tiny FDN Size

The optimization works with as few as **4 delay lines**. A traditional FDN typically uses 8-16 delay lines. This means:

- **4 delay lines** = ~4× fewer multiplications per sample
- **4 delay lines** = ~4× less memory
